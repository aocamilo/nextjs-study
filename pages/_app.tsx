import '../styles/globals.css'
import type { AppProps } from 'next/app'
import 'bootstrap/dist/css/bootstrap.css'
import styles from '../styles/Home.module.css'
import Link from 'next/link'

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <div className={styles.container}>
      <main className={styles.main}>
        <nav className="navbar navbar-expand-lg navbar-light bg-light position-absolute top-0 w-100 px-5">
          <div className="container-fluid">
            <Link href="/">
              <a className="navbar-brand">Milo</a>
            </Link>
            <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
              <span className="navbar-toggler-icon"></span>
            </button>
            <div className="collapse navbar-collapse" id="navbarNav">
              <ul className="navbar-nav">
                <li className="nav-item">
                  <Link href="/pokemon"><a className="nav-link active" aria-current="page">Pokemon</a></Link>
                </li>
                <li className="nav-item">
                  <Link href="/rickandmorty"><a className="nav-link active" aria-current="page">Rick And Morty</a></Link>
                </li>
                <li className="nav-item">
                  <Link href="/milos-blog/compound-components-in-react"><a className="nav-link active" aria-current="page">Compound Components</a></Link>
                </li>
                <li className="nav-item">
                  <Link href="/milos-blog/controlled-props-components-in-react"><a className="nav-link active" aria-current="page">Controlled Props</a></Link>
                </li>
              </ul>
            </div>
          </div>
        </nav>
        <Component {...pageProps} />
      </main>
      <footer className={styles.footer}>
        <p>Made By Camilo Arango - 2022 - ðŸ‡¨ðŸ‡´ </p>
      </footer>
    </div>
  )
}

export default MyApp
