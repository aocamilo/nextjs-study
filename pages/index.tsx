import type { NextPage } from "next";
import Head from "next/head";
import styles from "../styles/Home.module.css";

const Home: NextPage = () => {
  return (
    <div className={styles.container}>
      <Head>
        <title>Next Learn App</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <main className={styles.main}>
        <p data-testid="message">
          This is an app made to study and understand some concepts about
          Nextjs, SSR, SSG, CSR, and more... Please use the nav.
        </p>
      </main>
    </div>
  );
};

export default Home;
