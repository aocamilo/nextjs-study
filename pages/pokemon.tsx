import Image from "next/image";
import { useEffect, useState } from "react";
import Head from "next/head";
import Script from "next/script";
import Layout from "./components/layout";
import { usePokemonData } from "./hooks/usePokemonData";

interface LoaderProps {
  src: string;
  width: number;
  quality?: number;
}

export const loader = ({ src, width, quality }: LoaderProps) =>
  `${src}?w=${width}&q=${quality || 75}`;

/* 
  CSR (CLient Side Rendering)
  Next Build -> Generate Static HTML parts -> Fetch Data -> Populate the remaining HTML parts that depend on external data.
  Use this for pages that do not require SEO and are dependant on updated data.
*/

export interface PokemonAbility {
  ability: {
    name: {};
  };
}

export interface Pokemon {
  name: string;
  weight: string;
  sprites: {
    front_default: string;
  };
  abilities: PokemonAbility[];
}

export interface BasicPokemon {
  url: string;
}

const pokemon = () => {
  const [page, setPage] = useState(0);
  const { pokemones, loading } = usePokemonData(page);

  if (pokemones.length === 0 || loading) return <p>Loading...</p>;

  return (
    <Layout>
      <div className="container-fluid d-flex row align-items-center justify-content-center">
        <Head>
          <title>Pokedexapp - CSR</title>
        </Head>
        <Script
          src="https://connect.facebook.net/en_US/sdk.js"
          strategy="lazyOnload"
          onLoad={() =>
            console.log(`script loaded correctly, window.FB has been populated`)
          }
        />
        <h1 className="p-5">Pokedex</h1>
        {pokemones.map((pokemon: Pokemon) => (
          <div
            className="card d-flex col-3"
            key={pokemon.name}
            data-cy="pokemon-card"
          >
            <Image
              loader={loader}
              src={pokemon.sprites.front_default}
              alt={pokemon.name}
              width={96}
              height={96}
              layout="fixed"
            />
            <h2> {pokemon.name} </h2>
            <p>weight: {pokemon.weight}</p>
            <div className="d-flex flex-row m-2">
              {pokemon.abilities.map((ability: PokemonAbility) => (
                <span
                  className="badge bg-success me-1"
                  key={`${pokemon.name}-${ability.ability.name}`}
                >
                  {ability.ability.name}
                </span>
              ))}
            </div>
          </div>
        ))}
        <ul className="pagination justify-content-center p-5">
          <li className="page-item">
            <a
              className="page-link"
              onClick={() => setPage(Math.max(0, page - 1))}
            >
              Previous
            </a>
          </li>
          <li className="page-item disabled">
            <a className="page-link active" aria-current="page">
              {page + 1}
            </a>
          </li>
          <li className="page-item">
            <a className="page-link" onClick={() => setPage(page + 2)}>
              {page + 2}
            </a>
          </li>
          <li className="page-item">
            <a className="page-link" onClick={() => setPage(page + 3)}>
              {page + 3}
            </a>
          </li>
          <li className="page-item">
            <a className="page-link" onClick={() => setPage(page + 1)}>
              Next
            </a>
          </li>
        </ul>
      </div>
    </Layout>
  );
};

export default pokemon;
