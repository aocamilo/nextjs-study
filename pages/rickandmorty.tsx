import Head from "next/head";
import Image from "next/image";
import { loader } from "./pokemon";
import { useCallback, useState } from "react";

/*
  SSG only works on server-side and at build time, it can work either using external data or static data.
  if the pages requires external data it fetches the data and passes it as prop to the component. 
  It must be exported in the same page doc, like shown below. 
  Without data: Next Build -> HTML Generated
  With data: Next Build -> Fetch External Data -> HTML Generated
  Works like SSR in dev environment.

  getStaticProps: used to fetch external data, only runs server-side and works only at build time
  getStaticPaths: must be used with getStatic props, cannot be used with getServerSideProps, 
                  used if a page has Dynamic Routes to define a list of paths to be statically generated.

  note: Use only if the data does not change much and only initial fetch is required. CMS, Blog Posts, etc.
*/
// export async function getStaticProps() {
//   const res = await fetch('https://rickandmortyapi.com/api/character');
//   const characters = await res.json();

//   return {
//     props: {
//       characters: characters.results
//     }
//   }

// }

/* 
  SSR used when we need to fetch data at request time instead build time
  context parameter contains request specific parameters
  Page Request -> Next Build -> Fetch External Data -> HTML Generated.
  Optimal for SEO
  getServerSideProps: Runs at server-side only and is used at EVERY page request
*/

export async function getServerSideProps() {
  const res = await fetch("https://rickandmortyapi.com/api/character");
  const characters = await res.json();

  console.log(characters);

  if (characters.error) {
    return {
      props: {
        error: characters.error,
      },
    };
  }

  return {
    props: {
      characters: characters.results,
      next: characters.info.next,
    },
  };
}

export interface Character {
  name: string;
  image: string;
  id: number;
  gender: string;
  species: string;
  status: string;
}

interface Characters {
  characters: Character[];
  error?: string;
  next?: string;
}

const RickAndMorty = ({ characters, error, next }: Characters) => {
  const [filter, setFilter] = useState("");
  const [filteredChars, setFilteredChars] = useState(characters);
  const [runError, setRunError] = useState(null);

  const onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFilter(event.target.value.toLowerCase());

    if (event.target.value === "") {
      setFilteredChars(characters);
      return;
    }

    setFilteredChars(filterChars(filter));
  };

  const filterChars = useCallback(
    (f: string) => {
      return characters.filter((char) => char.name.toLowerCase().includes(f));
    },
    [filter]
  );

  const onResetFilter = () => {
    setFilteredChars(characters);
    setFilter("");
  };

  const onNextPage = async () => {
    console.log("called");

    if (next) {
      const res = await fetch(next);
      const resp = await res.json();

      if (resp.error) {
        setRunError(resp.error);
        return;
      }
    }
  };

  if (error || runError) return <>{error ?? runError}</>;

  if (!characters || characters.length === 0) return <>Loading...</>;

  return (
    <div className="container-fluid d-flex row align-items-center justify-content-center">
      <Head>
        <title>Rick And Morty - SSR / SSG</title>
      </Head>
      <div style={{ display: "flex", flexDirection: "row", width: "100%" }}>
        <input
          type="text"
          name="search"
          data-testid="filter-input-name"
          placeholder="Search Char"
          value={filter}
          onChange={onInputChange}
          style={{ position: "absolute", top: 70, width: "20%" }}
        />
        <button
          name="reset-filter"
          data-testid="reset-filter"
          onClick={onResetFilter}
          style={{
            marginTop: 50,
            position: "relative",
            top: 0,
            width: "20%",
          }}
        >
          Reset Filter
        </button>
        {next && (
          <button
            data-testid="fetch-next-page"
            name="fetch-more"
            onClick={onNextPage}
            style={{
              marginTop: 50,
              position: "relative",
              top: 0,
              width: "20%",
            }}
          >
            Fetch more
          </button>
        )}
      </div>

      {filteredChars.map(
        ({ name, image, id, gender, species, status }: Character) => (
          <div
            className="card d-flex col-3"
            key={`${id}-${name}`}
            style={{ marginTop: 10 }}
            data-testid="character"
          >
            <h2> {name} </h2>
            <Image
              loader={loader}
              src={image}
              alt={name}
              width={300}
              height={300}
              layout="fixed"
            />
            <p className="mt-5 mb-0 p-0">Gender: {gender}</p>
            <p className="m-0 p-0">Species: {species}</p>
            <p>Status: {status}</p>
          </div>
        )
      )}
    </div>
  );
};

export default RickAndMorty;
