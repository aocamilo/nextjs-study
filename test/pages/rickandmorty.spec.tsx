import { render, fireEvent, waitFor } from "@testing-library/react";
import RickAndMorty from "../../pages/rickandmorty";
import "@testing-library/jest-dom";
import rickAndMortyMock from "../rickAndMortyMock.json";
import { Character } from "../../pages/rickandmorty";

describe("When in the rick and morty app", () => {
  const originalFetch = window.fetch;
  let mockFetch: any;

  beforeEach(() => {
    mockFetch = jest.fn();
    window.fetch = mockFetch;
  });

  afterEach(() => {
    window.fetch = originalFetch;
  });

  function renderComponent() {
    const { getByTestId, getAllByTestId, asFragment, getByText } = render(
      <RickAndMorty
        characters={rickAndMortyMock as Character[]}
        next="https://rickandmortyapi.com/api/character?page=2"
      />
    );

    return {
      characters: getAllByTestId("character"),
      filterInput: getByTestId("filter-input-name") as HTMLInputElement,
      resetFilter: getByTestId("reset-filter"),
      fetchNext: getByTestId("fetch-next-page"),
      asFragment,
      getByText,
      getAllByTestId,
    };
  }

  it("It should match the snapshot", () => {
    const { asFragment } = renderComponent();

    expect(asFragment).toMatchSnapshot();
  });

  it("Should display 20 characters", () => {
    const { characters } = renderComponent();

    expect(characters.length).toBe(20);
  });

  it("Should set and clear filter correctly", () => {
    const { filterInput, resetFilter } = renderComponent();

    expect(filterInput.value).toBe("");

    fireEvent.change(filterInput, {
      target: { value: "morty" },
    });

    expect(filterInput.value).toBe("morty");

    fireEvent.click(resetFilter);

    expect(filterInput.value).toBe("");
  });

  it("Should call the API and display error correctly", async () => {
    mockFetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ error: "error..." }),
      });
    });

    const { fetchNext, getByText } = renderComponent();

    fireEvent.click(fetchNext);

    await waitFor(() => {
      expect(getByText("error...")).toBeInTheDocument();
    });
  });
});
