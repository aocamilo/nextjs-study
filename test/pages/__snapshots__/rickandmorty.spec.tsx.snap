// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When in the rick and morty app It should match the snapshot 1`] = `[Function]`;

exports[`When in the rick and morty app Should set and clear filter correctly 1`] = `[Function]`;
